src/privileged/vpnRecommender/api.js

```diff
@@ -0,0 +1,537 @@
+ */
+
+"use strict";
+
+/* global ExtensionAPI, ExtensionCommon, Preferences, PrivateBrowsingUtils */
+
+ChromeUtils.import("resource://gre/modules/Services.jsm");
+ChromeUtils.import("resource://gre/modules/XPCOMUtils.jsm");
+
+Cu.importGlobalProperties(["XMLHttpRequest", "URL", "URLSearchParams"]);
+
+XPCOMUtils.defineLazyModuleGetters(this, {
+  Preferences: "resource://gre/modules/Preferences.jsm",
+  setTimeout: "resource://gre/modules/Timer.jsm",
+  setInterval: "resource://gre/modules/Timer.jsm",
+  clearInterval: "resource://gre/modules/Timer.jsm",
+  PrivateBrowsingUtils: "resource://gre/modules/PrivateBrowsingUtils.jsm",
+  AddonManager: "resource://gre/modules/AddonManager.jsm",
+});
+
+const PREF_BRANCH = "extensions.vpn-recommendation-study-1_shield_mozilla_org";
+const DONT_SHOW_PREF = PREF_BRANCH + ".dontShowChecked";
+const NOTIFICATION_COUNT_PREF = PREF_BRANCH + ".notificationCount";
+const LAST_NOTIFICATION_PREF = PREF_BRANCH + ".lastNotification";
+const DEBUG_MODE_PREF = PREF_BRANCH + ".debug_mode";
+const TEST_PREF = PREF_BRANCH + ".started";
+const VARIATION_PREF = PREF_BRANCH + ".variation";
+
+const CP_SUCCESS_XHR_TIMEOUT = 3000;
+const CAPTIVE_PORTAL_URL = "http://detectportal.firefox.com/success.txt";
+const CP_SUCCESS_CHECK_INTERVAL = 10000;
+const CP_SUCCESS_MAX_CHECK_COUNT = 3;
+
+const TWENTY_FOUR_HOURS = 24 * 60 * 60 * 1000;
+
+const CATCH_ALL_TRIGGER_TIMER_MINUTES = 10;
+const MAX_NOTIFICATION_COUNT = 3;
+
+const VPN_LANDING_PAGE_URL = "https://premium.firefox.com/vpn/";
+const VPN_LANDING_PAGE_DEFAULT_PARAMS = {
+  "utm_source": "firefox-browser",
+  "utm_medium": "firefox-browser",
+  "utm_campaign": "shield_vpn",
+};
+
+const log = function(...args) {
+  if (!Preferences.get(DEBUG_MODE_PREF)) return;
+  console.log(...args);
+};
+
+const DOORHANGER_MESSAGES = {
+  "captive-portal": {
+    header: "It appears you are browsing on an unsecured wireless network.",
+    text: "Firefox has teamed up with ProtonVPN to provide you with a private and secure internet connect, no matter where you are.",
+  },
+  "privacy-hostname": {
+    header: "Make Firefox even more secure with ProtonVPN.",
+    text: "Firefox has teamed up with ProtonVPN to provide you with a private and secure internet connect, no matter where you are.",
+  },
+  "streaming-hostname": {
+    header: "Make Firefox even more secure with ProtonVPN.",
+    text: "Firefox has teamed up with ProtonVPN to provide you with a private and secure internet connect, no matter where you are.",
+  },
+  "catch-all": {
+    header: "Make Firefox even more secure with ProtonVPN.",
+    text: "Firefox has teamed up with ProtonVPN to provide you with a private and secure internet connect, no matter where you are.",
+  },
+};
+
+this.vpnRecommender = class extends ExtensionAPI {
+
+  getAPI(context) {
+    const that = this;
+
+    console.log("context", context);
+
+    this.jsms = {}; // workaround for not being able to use jsms with moz:// urls
```

Comment:
nit: are you using the member variable anywhere else? Could it be a local variable?

Reply:
Yes, now it's a local variable!